File: assetcore/services/shared/num_to_words_vi.py
# Copyright (c) 2026, AssetCore Team
"""Tiện ích đọc số tiền thành chữ tiếng Việt (VND).

Dùng chung cho mọi DocType cần hiển thị "số tiền bằng chữ"
(VD: Service Contract, Procurement, ...).
"""
from __future__ import annotations

_DIGITS = ("không", "một", "hai", "ba", "bốn", "năm", "sáu", "bảy", "tám", "chín")
_UNITS = ("", "nghìn", "triệu", "tỷ")


def _read_three_digits(number: int, full: bool) -> str:
    """Đọc một nhóm 3 chữ số.

    Args:
        number: giá trị 0..999.
        full: True nếu đây không phải nhóm cao nhất (cần đọc cả số 0 hàng trăm).
    """
    hundred, rem = divmod(number, 100)
    ten, unit = divmod(rem, 10)
    parts: list[str] = []

    if hundred > 0 or full:
        parts.append(_DIGITS[hundred])
        parts.append("trăm")

    if ten == 0:
        if unit > 0 and (hundred > 0 or full):
            parts.append("lẻ")
        if unit > 0:
            parts.append(_DIGITS[unit])
    elif ten == 1:
        parts.append("mười")
        if unit == 5:
            parts.append("lăm")
        elif unit > 0:
            parts.append(_DIGITS[unit])
    else:
        parts.append(_DIGITS[ten])
        parts.append("mươi")
        if unit == 1:
            parts.append("mốt")
        elif unit == 5:
            parts.append("lăm")
        elif unit > 0:
            parts.append(_DIGITS[unit])

    return " ".join(parts)


def num_to_words_vi(amount: float) -> str:
    """Đọc số tiền VND thành chữ tiếng Việt.

    Args:
        amount: số tiền (>= 0). Phần thập phân được làm tròn về số nguyên đồng.

    Returns:
        Chuỗi tiền bằng chữ, viết hoa chữ cái đầu, kết thúc bằng "đồng".
        Ví dụ: ``num_to_words_vi(1000000) == "Một triệu đồng"``.

    Raises:
        ValueError: khi ``amount`` âm.
    """
    if amount < 0:
        raise ValueError("Số tiền không được âm.")

    number = int(round(amount))
    if number == 0:
        return "Không đồng"

    groups: list[int] = []
    while number > 0:
        number, rem = divmod(number, 1000)
        groups.append(rem)

    segments: list[str] = []
    highest = len(groups) - 1
    for idx in range(highest, -1, -1):
        grp = groups[idx]
        if grp == 0:
            continue
        text = _read_three_digits(grp, full=(idx != highest))
        unit = _UNITS[idx]
        segments.append(f"{text} {unit}".strip())

    result = " ".join(segments).strip()
    return result[0].upper() + result[1:] + " đồng"

File: frontend/src/utils/numToWordsVi.ts
// Copyright (c) 2026, AssetCore Team
// Đọc số tiền VND thành chữ tiếng Việt — bản FE mirror chính xác BE
// `assetcore/services/shared/num_to_words_vi.py`. Dùng để preview trực tiếp
// trên form trước khi BE tính lại (BE vẫn là single source of truth khi lưu).

const DIGITS = ['không', 'một', 'hai', 'ba', 'bốn', 'năm', 'sáu', 'bảy', 'tám', 'chín']
const UNITS = ['', 'nghìn', 'triệu', 'tỷ']

/** Đọc một nhóm 3 chữ số (0..999). `full` = không phải nhóm cao nhất. */
function readThreeDigits(n: number, full: boolean): string {
  const hundred = Math.floor(n / 100)
  const rem = n % 100
  const ten = Math.floor(rem / 10)
  const unit = rem % 10
  const parts: string[] = []

  if (hundred > 0 || full) {
    parts.push(DIGITS[hundred], 'trăm')
  }

  if (ten === 0) {
    if (unit > 0 && (hundred > 0 || full)) parts.push('lẻ')
    if (unit > 0) parts.push(DIGITS[unit])
  } else if (ten === 1) {
    parts.push('mười')
    if (unit === 5) parts.push('lăm')
    else if (unit > 0) parts.push(DIGITS[unit])
  } else {
    parts.push(DIGITS[ten], 'mươi')
    if (unit === 1) parts.push('mốt')
    else if (unit === 5) parts.push('lăm')
    else if (unit > 0) parts.push(DIGITS[unit])
  }

  return parts.join(' ')
}

/**
 * Đọc số tiền VND thành chữ. Hỗ trợ đến hàng tỷ (giống BE).
 * Trả chuỗi rỗng nếu amount <= 0 hoặc không hợp lệ (BE lưu null tương ứng).
 */
export function numToWordsVi(amount: number | null | undefined): string {
  if (amount == null || !Number.isFinite(amount) || amount <= 0) return ''

  let number = Math.round(amount)
  const groups: number[] = []
  while (number > 0) {
    groups.push(number % 1000)
    number = Math.floor(number / 1000)
  }

  const segments: string[] = []
  const highest = groups.length - 1
  for (let idx = highest; idx >= 0; idx--) {
    const grp = groups[idx]
    if (grp === 0) continue
    const text = readThreeDigits(grp, idx !== highest)
    const unit = UNITS[idx] ?? ''
    segments.push(`${text} ${unit}`.trim())
  }

  const result = segments.join(' ').trim()
  return result.charAt(0).toUpperCase() + result.slice(1) + ' đồng'
}
